python_modules/dagster/dagster/_core/decorator_utils.py

```diff
@@ -87,6 +87,14 @@ def format_docstring_for_description(fn: Callable) -> Optional[str]:
 # When/if `StrictTypeGuard` is supported, we can drop `is_context_not_provided` since a False from
 # `is_context_provided` will be sufficient.
 def is_context_provided(
-    fn: Union[Callable[Concatenate[T, P], R], Callable[P, R]]
+    fn: Union[Callable[Concatenate[T, P], R], Callable[P, R]],
+    valid_names: Optional[Sequence[str]] = None,
 ) -> TypeGuard[Callable[Concatenate[T, P], R]]:
-    return len(get_function_params(fn)) >= 1
+    params = get_function_params(fn)
+    if valid_names is None:
+        return len(params) >= 1
+    else:
+        valid_name_permutations = [
+            x for name in valid_names for x in get_valid_name_permutations(name)
+        ]
+        return len(params) >= 1 and params[0].name in valid_name_permutations
```

Comment:
This doesn't really fix one of the problems here, which is that the function here is named inaccurately. What it means now is that "return true if these is more than one argument of any name. If valid_names is set the function then returns true if param name is raw underscore. or one of the valid names, or one the valid names preceded by a "_" or suffixed with "_".

This is an extremely complicated semantic and seems like a lot for a generic function. In particular that it takes multiple `valid_names` makes this particularly problematic and hard to understand.

I recommend that we instead change the name of this function from `is_context_provided` to `has_at_least_one_parameter` so that the name is _accurate_. Then all the callsites to this will make more sense where they have errors conditions that multiple arguments are illegal.

Reply:
I agree a generic function should have a different name. It's unclear to me from this though whether you're asking for:

- Two functions to be preserved with the `decorator_utils` one renamed
- The consolidated function to be renamed

Reply:
I think for now have two sep functions with two names for clarity. Changing the naming constraints is an entire other matter that would involve a deprecation cycle, etc. Let's just make the current state of the world more clear.